xdsl/traits.py

```diff
@@ -342,3 +343,20 @@ def get_callable_region(cls, op: Operation) -> Region:
         Returns the body of the operation
         """
         raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class HasCanonicalisationPatternsTrait(OpTrait):
+    """
+    Provides the rewrite passes to canonicalize an operation.
+
+    Each pass must have the trait's op as root.
+    """
+
+    def verify(self, op: Operation) -> None:
+        return
+
+    @classmethod
+    @abc.abstractmethod
+    def get_canonicalization_patterns(cls) -> tuple[RewritePattern, ...]:
+        raise NotImplementedError()
```

Comment:
This should rather be an `...`, or an assert.
`NotImplementedError` has a different meaning here.

Reply:
I'm a little confused, it's what we use throughout the framework for abstract methods. How do you decide between `...` and `raise NotImplementedError()`?